### PR TITLE
perf(store): run change detection once for all selectors when asynchronous action has been completed

### PR DESCRIPTION
**Pull request by arturovt**

### CODE REVIEW

File: packages/store/tests/zone.spec.ts [hunk 1]
```diff
@@ -25,53 +25,10 @@ describe('zone', () => {
     }
   }
 
-  describe('[store.select]', () => {
```
Comment (arturovt) [inline packages/store/tests/zone.spec.ts:28]: Removed it since it duplicates the below test.

**Review by splincode — APPROVED** [review]

File: packages/store/src/internal/state-stream.ts [hunk 2]
```diff
@@ -3,13 +3,19 @@ import { BehaviorSubject } from 'rxjs';
...
 @Injectable()
 export class StateStream extends BehaviorSubject<PlainObject> {
-  constructor() {
+  constructor(private _ngxsExecutionStrategy: InternalNgxsExecutionStrategy) {
     super({});
   }
+
+  next(value: PlainObject): void {
+    this._ngxsExecutionStrategy.leave(() => super.next(value));
+  }
```
Comment (markwhitfeld) [inline packages/store/src/internal/state-stream.ts:20]: I think that this could have unintended side effects for other aspects of the state management (like plugins etc.).
I think it would be a better approach if `Store` had a derived state stream that calls the `leave` to change the execution context, and then is shared using the `shareReplay` operator.
All selects would then use this, and it would call `leave` only once for any state change across all active selectors. What do you think @arturovt?
This would be used by the `select` and `subscribe` methods of the `Store`.

**Review by markwhitfeld — CHANGES_REQUESTED** [review]

File: packages/store/src/internal/state-stream.ts (hunk 2 above)
Comment (arturovt) [reply packages/store/src/internal/state-stream.ts]: @markwhitfeld I've made some changes, could you have a look. This is the only approach that doesn't break unit tests, using `shareReplay` breaks a lot of tests, I didn't dig why 😄

File: packages/store/src/store.ts [hunk 3]
```diff
@@ -87,7 +98,9 @@ export class Store {
    * Allow the user to subscribe to the root of the state
    */
   subscribe(fn?: (value: any) => void): Subscription {
-    return this._stateStream.pipe(leaveNgxs(this._internalExecutionStrategy)).subscribe(fn);
+    return this._selectableStateStream
+      .pipe(leaveNgxs(this._internalExecutionStrategy))
```
Comment (markwhitfeld) [inline packages/store/src/store.ts:102]: Is this still necessary since we are using the new shared stream?

File: packages/store/tests/issues/issue-933-selectors-causing-ticks.spec.ts [hunk 4]
```diff
@@ -53,7 +53,7 @@ describe('Selectors within templates causing ticks (https://github.com/ngxs/stor
   class TestModule {}
 
   it(
-    'should run change detection whenever the selector emits after asynchronous action has been completed',
+    'should not run change detection whenever the selector emits after asynchronous action has been completed',
```
Comment (markwhitfeld) [inline packages/store/tests/issues/issue-933-selectors-causing-ticks.spec.ts:56]: Potentially better to say "should run change detection once for all selectors when acting action has been completed"

File: packages/store/src/store.ts (hunk 3 above)
Comment (arturovt) [reply packages/store/src/store.ts:102]: This is necessary in case the Store class is subscribed outside of the Angular zone.

Comment (markwhitfeld) [reply packages/store/src/store.ts:102]: I thought that the new `_selectableStateStream` would handle ensuring that the subscribers are notified in the correct zone?
Or is this for the case where we are reading the current value that already exists in the stream?
If this is the case, then would we not also have to do this if the `store.select` is subscribed to outside of the Angular Zone?

File: packages/store/src/store.ts [hunk 5]
```diff
@@ -25,6 +35,7 @@ export class Store {
     @Inject(INITIAL_STATE_TOKEN)
     initialStateValue: any
   ) {
+    (this._selectableStateStream as ConnectableObservable<PlainObject>).connect();
```
Comment (markwhitfeld) [inline packages/store/src/store.ts:38]: I think that there may be an issue with the explicit `connect()` where, if the subscriber count reaches 0, then the stream will disconnect and I think it won't reconnect. Potentially we would want to add a `refCount()` after the `publishReplay(1)` pipe in the observable declaration and get rid of the manual `connect()`.
We definitely need to have a test that checks a subscription to the select and then subscription (taking the subscriptions to zero) and then after a state change, subscribing to a select again and checking it gets the latest state as well as further updates.
This article goes into great depth to explain the differences:
https://itnext.io/the-magic-of-rxjs-sharing-operators-and-their-differences-3a03d699d255

File: packages/store/src/store.ts (hunk 3 above)
Comment (markwhitfeld) [reply packages/store/src/store.ts:102]: I thought that the new `_selectableStateStream` would handle ensuring that the subscribers are notified in the correct zone? But maybe there would be no guarantee for new subscriptions where we first receive the current value that already exists in the stream?
If this is the case, then would we not also have to do this if the `store.select` is subscribed to outside of the Angular Zone?

**Review by markwhitfeld — CHANGES_REQUESTED** [review]

File: packages/store/src/store.ts (hunk 3 above)
Comment (arturovt) [reply packages/store/src/store.ts:109]: The `_selectableStateStream` cannot ensure that completely. Given this example:
```ts
constructor(store: Store, ngZone: NgZone) {
  ngZone.runOutsideAngular(() => {
    setTimeout(() => {
      store.subscribe(() => console.log(Zone.current)); // will log angular
    }, 3000);
  });
}
```
Otherwise, it'll log `<root>`.

**Comment by markwhitfeld** [comment]
@arturovt is this still in draft? 
Is there anything you need from me to get this ready?

**Comment by arturovt** [comment]
Glad you're back! No, I've made is as a draft initially, but this is ready now to be merged (if all requirements are passed).

**Review by markwhitfeld — APPROVED** [review]
I'm happy with this. Just need to fix the failing bundlesizes ;-)
Might be worth merging master into this branch before doing this too.

**Comment by arturovt** [comment]
Updated, hope is passes, this bundlesize is such annoying guy... 😸

**Comment by markwhitfeld** [comment]
I couldn't agree more!!!
https://github.com/ngxs/store/issues/1764
